### PR TITLE
Address issues with tracking dirty & pending states

### PR DESCRIPTION
**Pull request by vladsud**
Splitting last commit from https://github.com/microsoft/FluidFramework/pull/11811 into its own PR.
We are not correctly handling dirty flag transitions when applying stashed ops.

Opened the following bugs to track issues discovered in Pending State manager and dirty flag management:
- https://dev.azure.com/fluidframework/internal/_workitems/edit/1832
- https://dev.azure.com/fluidframework/internal/_workitems/edit/1833

### CODE REVIEW

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 1]
```diff
@@ -233,7 +233,11 @@ export class PendingStateManager implements IDisposable {
 
             // then we push onto pendingStates which will cause PendingStateManager to resubmit when we connect
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            this.pendingStates.push(this.initialStates.shift()!);
+            const message = this.initialStates.shift()!;
+            this.pendingStates.push(message);
+            if (message.type === "message") {
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:238]: #nit technically `message` is a pending state which may be of type message, so the const would probably be better named as `firstPendingState` or something (naming is hard)

**Review by andre4i — APPROVED** [review]
👍

File: packages/runtime/container-runtime/src/pendingStateManager.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:238]: Fixing

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -2902,6 +2917,13 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         // we may not have seen every sequence number (because of system ops) so apply everything once we
         // don't have any more saved ops
         await this.pendingStateManager.applyStashedOpsAt();
+
+        // If it's not the case, we should take it into account when calculating dirty state.
+        assert(this.context.attachState === AttachState.Attached,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2922]: Isn't it possible for `processSavedOps` to be called when rehydrating a detached container?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -2529,7 +2532,19 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    private hasPendingMessages() {
+        return this.pendingStateManager.hasPendingMessages() || !this.batchManager.empty;
+    }
+
     private updateDocumentDirtyState(dirty: boolean) {
+        if (this.attachState !== AttachState.Attached) {
+            assert(dirty, "Non-attached container is dirty");
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2541]: nit: "Non-attached container is _not_ dirty"

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2922]: Today's design assumes on of the two:
1. Detached container: there are no ops at all. We serialize full state (snapshot) and load from it.
2. Attached container: we serialize base snapshot, sequenced ops on top of it and local (stashed) ops.

Check in constructor ensures this API is called only for case # 2 (we have pending state).

If design changes in the future, we will adjust accordingly

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2922]: That makes sense. I did not realize that detached container won't have any stashed ops :-).

**Pull request merged**